main/src/modules/reactiveLayer/sensationManager/src/opcSensation.cpp

```diff
@@ -133,16 +137,78 @@ Bottle OpcSensation::handleEntities()
             else if (entity->entity_type() == "object" && dynamic_cast<Object*>(entity)->m_present == 1.0) {  // Known entities and present!
                 known_obj = true;
                 addToEntityList(k_entities, entity->entity_type(), entity->name());
+                //Handle red balls
+                if (entity->name() == "red_ball"){
+                    yDebug("Found a redBall object");
+                    if (!Network::isConnected("/pf3dTracker/data:o",("/"+moduleName+"/pf3dTracker:i").c_str())){
+                        yDebug("RedBall Port not connected...");
+                        dynamic_cast<Object*>(entity)->m_present = 0.0;
+                    }else{
+                        yDebug("RedBall Port connected...\nUpdateing position...");
+                        //update position
+                        Bottle *bot = pf3dTrackerPort.read();
+                        yDebug("RedBall Port readed");
+                        Object* obj1 = iCub->opc->addOrRetrieveEntity<Object>("red_ball");
```

Comment:
No need to do this, you already have the entity you want. Just cast it to an `Object`:
`Object* obj1 = dynamic_cast<Object*>(entity)`